desktop/cmp/treemap/TreeMapModel.js

```diff
@@ -0,0 +1,396 @@
+                        it.colorValue = this.normalizeToRange(heatValue, minPosHeat, midPosHeat, 0.6, 0.8);
+                    }
+                } else if (colorMode === 'linear') {
+                    it.colorValue = this.normalizeToRange(heatValue, minPosHeat, maxPosHeat, 0.6, 1);
+                }
+            } else if (heatValue < 0) {
+                // Normalize negative values between 0-0.4
+                const absHeatValue = Math.abs(heatValue);
+
+                if (colorMode === 'balanced' && posHeatValues.length > 2) {
+                    if (absHeatValue >= midNegHeat) {
+                        it.colorValue = this.normalizeToRange(absHeatValue, maxNegHeat, midNegHeat, 0, 0.2);
+                    } else {
+                        it.colorValue = this.normalizeToRange(absHeatValue, midNegHeat, minNegHeat, 0.2, 0.4);
+                    }
+                } else if (colorMode === 'linear') {
+                    it.colorValue = this.normalizeToRange(absHeatValue, maxNegHeat, minNegHeat, 0, 0.4);
+                }
+            } else {
+                it.colorValue = 0.5; // Exactly zero
+            }
+        });
+
+        return data;
+    }
+
+    /**
+     * Takes a value, calculates its normalized (0-1) value within a specified range.
+     * If a destination range is provided, returns that value transposed to within that range.
+     */
+    normalizeToRange(value, fromMin, fromMax, toMin, toMax) {
+        const fromRange = (fromMax - fromMin),
+            toRange = (toMax - toMin);
+
+        if (isFinite(toRange)) {
+            // Return value transposed to destination range
+            return (((value - fromMin) * toRange) / fromRange) + toMin;
+        } else {
+            // Return value between 0-1
+            return (value - fromMin) / fromRange;
+        }
+    }
+
+    //----------------------
+    // Expand / Collapse
+    //----------------------
+    nodeIsExpanded(xhTreePath) {
+        if (isEmpty(this.expandState)) return false;
+        return get(this.expandState, xhTreePath, false);
+    }
+
+    toggleNodeExpanded(xhTreePath) {
+        const {gridModel} = this,
+            expandState = {...gridModel.expandState};
+
+        if (get(expandState, xhTreePath)) {
+            unset(expandState, xhTreePath);
+        } else {
+            set(expandState, xhTreePath, true);
+        }
+
+        gridModel.setExpandState(expandState);
+    }
+
+    //----------------------
+    // Click handling
+    //----------------------
+    defaultOnClick = (record, e) => {
+        if (!this.gridModel) return;
+
+        // Select nodes in grid
+        const {selModel} = this.gridModel;
+        if (selModel.mode === 'disabled') return;
+
+        const multiSelect = selModel.mode === 'multiple' && e.shiftKey;
+        selModel.select(record, !multiSelect);
+    };
+
+    defaultOnDoubleClick = (record) => {
+        if (!this.gridModel || !this.gridModel.treeMode || !record.children) return;
```

Comment:
Looks like `children` is an empty array for leaf nodes. Let's check `isEmpty(record.children)` instead to avoid triggering downstream work for the gridModel, etc.